snapshotter/demux/cache/snapshotter_cache.go

```diff
@@ -25,40 +25,41 @@ import (
 // SnapshotterCache implements a read, write protected cache mechanism
 // for keyed snapshotters.
 type SnapshotterCache struct {
-	mutex        *sync.Mutex
+	mutex        *sync.RWMutex
 	snapshotters map[string]snapshots.Snapshotter
 }
 
 // NewSnapshotterCache creates a new instance with an empty cache.
 func NewSnapshotterCache() *SnapshotterCache {
-	return &SnapshotterCache{&sync.Mutex{}, make(map[string]snapshots.Snapshotter)}
+	return &SnapshotterCache{&sync.RWMutex{}, make(map[string]snapshots.Snapshotter)}
 }
 
 // Get fetches and caches the snapshotter for a given key.
 func (cache *SnapshotterCache) Get(ctx context.Context, key string, fetch SnapshotterProvider) (snapshots.Snapshotter, error) {
+	cache.mutex.RLock()
 	snapshotter, ok := cache.snapshotters[key]
+	cache.mutex.RUnlock()
 
 	if !ok {
 		cache.mutex.Lock()
 		defer cache.mutex.Unlock()
-
-		snapshotter, ok = cache.snapshotters[key]
-		if !ok {
-			newSnapshotter, err := fetch(ctx, key)
-			if err != nil {
-				return nil, err
-			}
-
-			cache.snapshotters[key] = newSnapshotter
-			snapshotter = newSnapshotter
```

Comment:
I think we want to keep the double check. There is still the chance we leak resources if two threads have a cache miss and end up populating the cache.

Reply:
Ahh I remember now: https://github.com/firecracker-microvm/firecracker-containerd/pull/588#discussion_r822192190 . I'll revert to include the double check 🙂